feat(funnels): speed up funnel queries by preventing properties being passed across shards

yakkomajuri opened the pull request:
## Problem

#13922 

## Changes

Derive the funnel steps in the innermost subquery in order to prevent sending full properties across shards. This can speed up queries by up to 3-5x (estimate based on the fact that the slower the query today the greater the gains). 

The current query will send data in the order of megabytes to the other shard, whereas the previous query would send gigabytes and gigabytes (see e.g. first pic).

Benchmarks:

<img width="800" alt="Screenshot 2023-01-26 at 11 23 01" src="https://user-images.githubusercontent.com/38760734/214860106-b9f871f2-679d-453b-ae09-2bd3967b15ee.png">

<details>

<summary><b>See more</b></summary>

</br>
<img width="1508" alt="Screenshot 2023-01-25 at 14 44 26" src="https://user-images.githubusercontent.com/38760734/214856640-f02b8bf7-f5c5-4b3b-8fb9-568c5120832d.png">
<img width="1499" alt="Screenshot 2023-01-25 at 14 44 36" src="https://user-images.githubusercontent.com/38760734/214856690-8b488224-ca3f-4dbd-8bcc-8dce13c72436.png">
</details>

Note that I chose the path of least resistance to implement this. I think there's a refactor in order to simplify a lot of the logic (which is very messy) but I decided not to do that here. I kept things mostly as they are, with few changes.


A bunch of stale snapshots got deleted as a result of my testing.

## How did you test this code?

Existing tests for funnels are all passing. Also tested manually.

Marked it as a feat because of the perf improvement but from the perspective of the query results it's just a refactor - no behavior should change.

yakkomajuri committed:
feat(funnels): speed up funnel queries by preventing properties being passed across shards

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

yakkomajuri commented:
Well.... Seems all's good for the "traditional funnel" and f'd for everything else :D

yakkomajuri committed:
testing

yakkomajuri committed:
Merge branch 'funnel-props' of https://github.com/PostHog/posthog into funnel-props

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

yakkomajuri commented:
Seems only unordered is the problem (besides this one tricky thing with exclusions that I know how to fix), looking into it

yakkomajuri committed:
new approach

yakkomajuri committed:
Merge branch 'funnel-props' of https://github.com/PostHog/posthog into funnel-props

yakkomajuri committed:
merge master

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

yakkomajuri committed:
remove more unnecessary fields

yakkomajuri committed:
Merge branch 'funnel-props-playground' into funnel-props
Remove more unnecessary fields

yakkomajuri committed:
update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

yakkomajuri commented:
There are more columns that can be removed from the select and potentially more de-nesting. Will look into it tomorrow

yakkomajuri committed:
remove more cols and see :)

yakkomajuri committed:
remove event_columns_to_query

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

macobo commented on posthog/queries/funnels/funnel_event_query.py:
```diff
@@ -112,17 +86,35 @@ def get_query(self, entities=None, entity_name="events", skip_entity_filter=Fals
+            + "{step_filter}"
+        )
+
+        query = f"""
```
Q: Why split the overall thing up? IMO not adding too much clarity here.

yakkomajuri replied:
We spoke about this on a call - essentially there are ordering requirements with the functions building this query (as they have side effects) and this was the cleanest way to build things now without a large refactor

macobo commented on posthog/queries/funnels/funnel_event_query.py:
```diff
@@ -112,17 +86,35 @@ def get_query(self, entities=None, entity_name="events", skip_entity_filter=Fals
-        query = f"""
-            SELECT {', '.join(_fields)} FROM events {self.EVENT_TABLE_ALIAS}
+        select_fragment = (
+            f"SELECT {', '.join(_fields)}" + "{extra_select_fields}" + f" FROM events {self.EVENT_TABLE_ALIAS}"
```
You can do the following in f-strings: `{{extra_select_fields}}` if really needed, though the logic behind it is sus.

yakkomajuri replied:
updated to this

yakkomajuri committed:
update query building

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

yakkomajuri committed:
Merge branch 'master' into funnel-props

github-actions[bot] committed:
Update snapshots

yakkomajuri commented:
Ok this is ready to go @macobo

yakkomajuri commented on ee/clickhouse/views/test/__snapshots__/test_clickhouse_experiments.ambr:
```diff
@@ -157,36 +150,29 @@
-                   (SELECT aggregation_target,
-                           timestamp,
+                   (SELECT e.timestamp as timestamp,
+                           pdi.person_id as aggregation_target,
+                           pdi.person_id as person_id ,
```
we certainly don't need this :D

Kinda don't wanna touch this PR anymore though so will fix in a follow up if OK

github-actions[bot] committed:
Update snapshots

github-actions[bot] committed:
Update snapshots

macobo commented on posthog/queries/funnels/base.py:
```diff
@@ -398,9 +397,7 @@ def _get_inner_event_query(
-        parsed_extra_fields = f", {', '.join(extra_fields)}" if extra_fields else ""
-
-        event_query, params = FunnelEventQuery(
+        inner_query, params = FunnelEventQuery(
```
Q: Should we name this `events_query`?

macobo commented on posthog/queries/funnels/base.py:
```diff
@@ -426,29 +423,27 @@ def _get_inner_event_query(
         if self._filter.breakdown:
             if self._filter.breakdown_type == "cohort":
                 extra_join = self._get_cohort_breakdown_join()
+
```
Nit: Let's remove this empty line
```suggestion
```

macobo commented on posthog/queries/funnels/base.py:
```diff
@@ -426,29 +423,27 @@ def _get_inner_event_query(
-            event_query=event_query,
+        extra_select_fields = f", {', '.join(all_step_cols)}" if all_step_cols else ""
+
+        inner_query = inner_query.format(
```
Q: Should we name this `events_query` or `funnel_events_query`? inner_query had me going in cycles since it was ambigious.

macobo commented:
Amazing work!

yakkomajuri commented:
I'll get this in because every change I've been making seems to be a bit of a struggle with checks failing due to API limits and what not. Want to get this out of sight out of mind :D Will put a PR right away addressing the review points though @macobo
